ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -705,52 +751,64 @@ int getSize() {
       DataOutputStream dataOutputStream = new DataOutputStream(outStream);
 
       /* Write the number of entries */
-      dataOutputStream.writeInt(blobReadOptionsList.size());
+      dataOutputStream.writeInt(items.size());
 
       /* Write all the blobReadOptions */
-      for (BlobReadOptions blobReadOptions : blobReadOptionsList) {
-        dataOutputStream.write(blobReadOptions.toBytes());
+      for (HardDeletePersistItem item : items) {
+        dataOutputStream.write(item.blobReadOptions.toBytes());
       }
 
       /* Write all the messageStoreRecoveryInfos */
-      for (byte[] recoveryInfo : messageStoreRecoveryInfoList) {
+      for (HardDeletePersistItem item : items) {
         /* First write the size of the recoveryInfo */
-        dataOutputStream.writeInt(recoveryInfo.length);
+        dataOutputStream.writeInt(item.messagesStoreRecoveryInfo.length);
 
         /* Now, write the recoveryInfo */
-        dataOutputStream.write(recoveryInfo);
+        dataOutputStream.write(item.messagesStoreRecoveryInfo);
       }
 
       return outStream.toByteArray();
     }
 
     /**
-     * Prunes entries in the range from the start up to, but excluding, the entry with the passed in key.
+     * Prunes entries in the range from the start up to, but excluding, the entry with the passed in token.
+     * Token passed to this method has to be a indexed based one.
      */
-    void pruneTill(StoreKey storeKey) {
-      Iterator<BlobReadOptions> blobReadOptionsListIterator = blobReadOptionsList.iterator();
-      Iterator<byte[]> messageStoreRecoveryListIterator = messageStoreRecoveryInfoList.iterator();
-      while (blobReadOptionsListIterator.hasNext()) {
-      /* Note: In the off chance that there are multiple presence of the same key in this range due to prior software
-         bugs, note that this method prunes only till the first occurrence of the key. If it so happens that a
-         later occurrence is the one really associated with this token, it does not affect the safety.
-         Persisting more than what is required is okay as hard deleting a blob is an idempotent operation. */
-        messageStoreRecoveryListIterator.next();
-        if (blobReadOptionsListIterator.next().getMessageInfo().getStoreKey().equals(storeKey)) {
-          break;
+    void pruneTill(StoreFindToken token) {
+      Iterator<HardDeletePersistItem> itemsIterator = items.iterator();
+      while (itemsIterator.hasNext()) {
+        HardDeletePersistItem item = itemsIterator.next();
+        if (item.startTokenForBlobReadOptions.getType() == FindTokenType.Uninitialized) {
```

Comment:
Could you explain a little bit about this case?

Reply:
The very first time when we starting scanning through the log file, we are using a Uninitialized FindToken. And since it's at the very first beginning of a log file, we can assume they are always prior to any IndexBased FindToken.